amy/workshops/util.py

```diff
@@ -994,6 +996,51 @@ def assign(request, obj, person_id):
         raise Http404("No person found matching the query.")
 
 
+def archive_least_recent_active_consents(object_a, object_b, base_obj):
+    """
+    There is a unique database constraint on consents that only allows
+    (person, term) when archived_at is null.
+
+    This method archives one of the two active terms so
+    that the combine merge method will be successful.
+    """
+    consents = Consent.objects.filter(person__in=[object_a, object_b])
+    # Identify and group the active consents by term id
+    active_consents_by_term_id = defaultdict(list)
+    for consent in consents:
+        if consent.archived_at is None:
+            active_consents_by_term_id[consent.term_id].append(consent)
+
+    # archive least recent active consents
+    consents_to_archive = []
+    consents_to_recreate = []
+    for term_consents in active_consents_by_term_id.values():
+        if len(term_consents) < 2:
+            continue
+        consent_a, consent_b = term_consents[0], term_consents[1]
+        if consent_a.created_at < consent_b.created_at:
+            consents_to_archive.append(consent_a)
+        elif consent_b.created_at < consent_a.created_at:
+            consents_to_archive.append(consent_b)
+        else:
+            # If they were created at the same time rather than being
+            # nondeterministic archive both and when the user logs in again
+            # they can consent to the term once more.
+            consents_to_archive.append(consent_a)
+            consents_to_archive.append(consent_b)
+            consents_to_recreate.append(
```

Comment:
If both consents are archived, why only one is being recreated?

If only one object is being created, then `consents_to_recreate` doesn't have to be a list, and you don't need to use `bulk_create` down below.

Reply:
Okay, I think I understand why only one consent is being recreated. It will appear for the `base_obj` user, who is the result of a merge.

Reply:
@pbanaszkiewicz that's correct. I can add a better comment there. Basically, if I couldn't determine which consent to pick I archived both and required the new merged user to reconsent. I can also just go with the base_obj's consent instead though. Did you have feelings about that?